messaging/src/test/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvokerTest.java

```diff
@@ -139,6 +139,110 @@ void handleMethodHandlesEventJustFine() throws Exception {
+        verify(handler).handle(TEST_EVENT);
+
+        verify(queue, never()).enqueueIfPresent(eq(TEST_SEQUENCE_ID), any());
+
+        verify(queue, never()).enqueue(eq(TEST_SEQUENCE_ID), any());
+        verifyNoInteractions(transactionManager);
+    }
+
+    @Test
+    void handleMethodHandlesEventJustFineWithCacheWhenDlqNotEmpty() throws Exception {
+        setTestSubject(createTestSubject(b -> b.cacheEnabled(true)));
+        doReturn(1L).when(queue).amountOfSequences();
+        GenericDeadLetter.clock = Clock.fixed(Instant.now(), ZoneId.systemDefault());
+
+        when(queue.enqueueIfPresent(any(), any())).thenReturn(false);
+
+        testSubject.handle(TEST_EVENT, Segment.ROOT_SEGMENT);
+
+        verify(sequencingPolicy, times(2)).getSequenceIdentifierFor(TEST_EVENT);
+        verify(handler).handle(TEST_EVENT);
+
+        verify(queue, times(1)).enqueueIfPresent(eq(TEST_SEQUENCE_ID), any());
+
+        verify(queue, never()).enqueue(eq(TEST_SEQUENCE_ID), any());
+        verifyNoInteractions(transactionManager);
+    }
+
+    @Test
+    void handleMethodHandlesEventJustFineWithCacheWhenDlqNotEmptyKeepsTrackNotInDlq() throws Exception {
+        setTestSubject(createTestSubject(b -> b.cacheEnabled(true)));
+        doReturn(1L).when(queue).amountOfSequences();
+        GenericDeadLetter.clock = Clock.fixed(Instant.now(), ZoneId.systemDefault());
+
+        when(queue.enqueueIfPresent(any(), any())).thenReturn(false);
+
+        EventMessage<?> eventMessageOne = createEvent("foo", 2);
+        EventMessage<?> eventMessageTwo = createEvent("bar", 2);
+        EventMessage<?> eventMessageThree = createEvent("foo", 3);
+
+        testSubject.handle(eventMessageOne, Segment.ROOT_SEGMENT);
+        testSubject.handle(eventMessageTwo, Segment.ROOT_SEGMENT);
+        testSubject.handle(eventMessageThree, Segment.ROOT_SEGMENT);
+
+        verify(queue, times(1)).enqueueIfPresent(eq("foo"), any());
+        verify(queue, times(1)).enqueueIfPresent(eq("bar"), any());
+
+        verify(queue, never()).enqueue(eq(TEST_SEQUENCE_ID), any());
+        verifyNoInteractions(transactionManager);
+    }
+
+    @Test
+    void handleMethodHandlesEventJustFineWithCacheWhenDlqNotEmptyAndRespectCacheSize() throws Exception {
+        setTestSubject(createTestSubject(b -> b
+                .cacheEnabled(true)
+                .cacheSize(1)));
+        doReturn(1L).when(queue).amountOfSequences();
+        GenericDeadLetter.clock = Clock.fixed(Instant.now(), ZoneId.systemDefault());
+
+        when(queue.enqueueIfPresent(any(), any())).thenReturn(false);
+
+        DomainEventMessage<?> eventMessageOne = createEvent("foo", 2);
+        DomainEventMessage<?> eventMessageTwo = createEvent("bar", 2);
+        DomainEventMessage<?> eventMessageThree = nextMessage(eventMessageOne);
+
+        testSubject.handle(eventMessageOne, Segment.ROOT_SEGMENT);
+        testSubject.handle(eventMessageTwo, Segment.ROOT_SEGMENT);
```

Comment:
Nit: perhaps add a line that states that once the cache size is reached, the oldest entry is removed.

Reply:
Added